analytical_engine/apps/clustering/transitivity.h

```diff
@@ -48,12 +48,13 @@ class Transitivity
 
     messages.InitChannels(thread_num());
     ctx.stage = 0;
-    ForEach(inner_vertices, [&messages, &frag, &ctx](int tid, vertex_t v) {
-      ctx.global_degree[v] =
-          frag.GetLocalOutDegree(v) + frag.GetLocalInDegree(v);
-      messages.SendMsgThroughEdges<fragment_t, int>(frag, v,
-                                                    ctx.global_degree[v], tid);
-    });
+    ForEach(inner_vertices.begin(), inner_vertices.end(),
+            [&messages, &frag, &ctx](int tid, vertex_t v) {
+              ctx.global_degree[v] =
+                  frag.GetLocalOutDegree(v) + frag.GetLocalInDegree(v);
+              messages.SendMsgThroughEdges<fragment_t, int>(
+                  frag, v, ctx.global_degree[v], tid);
+            });
```

Comment:
I'm wondering why those two `ForEach` yeilds different semantics here, isn't `ForEach(inner_vertices)` equivalent to `ForEach(inner_vertices.begin(), inner_vertices.end())`? Just looks a bit counterintuitive ..

Reply:
`ForEach(inner_vertices.begin(), inner_vertices.end())` is more general than `ForEach(inner_vertices)`. `ForEach(inner_vertices)` are a series of specified template `ForEach` which only work for vertex_array_t defined in`libgrape-lite`, and `ForEach(inner_vertices.begin(), inner_vertices.end())` can work for any self-defined vertex_array_t    which defines iterators, such as `UnionVertexRange` in `ArrowFlattenedFragment`.

Reply:
> `ForEach(inner_vertices.begin(), inner_vertices.end())` is more general than `ForEach(inner_vertices)`. `ForEach(inner_vertices)` are a series of specified template `ForEach` which only work for vertex_array_t defined in`libgrape-lite`, and `ForEach(inner_vertices.begin(), inner_vertices.end())` can work for any self-defined vertex_array_t which defines iterators, such as `UnionVertexRange` in `ArrowFlattenedFragment`.

Fair enough. Thanks for clearifying :heart: